In [3]:
!pip install git+https://github.com/m-bain/whisperx.git --upgrade

  Cloning https://github.com/m-bain/whisperx.git to /tmp/pip-req-build-m96_mhg3
  Running command git clone --filter=blob:none --quiet https://github.com/m-bain/whisperx.git /tmp/pip-req-build-m96_mhg3
  Resolved https://github.com/m-bain/whisperx.git to commit dbeb8617f298bb4b5847d771bfb600379255c860
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 10.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.7/7.7 MB 32.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 200.2/200.2 kB 27.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.6/44.6 kB 7.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.0/302.0 kB 33.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 51.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.5/79.5 kB 11.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 153.4/153.4 MB 1.9 MB/s eta 0:00:00
     ━━━━━

In [1]:
from google.colab import files

uploaded = files.upload()
audio_file = list(uploaded.keys())[0]

Saving audio1.wav to audio1.wav


# **Transcripción del audio**

In [4]:
import whisperx
import gc
import torch

device = "cuda"
audio_file = "audio1.wav"
batch_size = 16 # reduce if low on GPU mem
compute_type = "float16" # change to "int8" if low on GPU mem (may reduce accuracy)

# 1. Transcribe with original whisper (batched)
model = whisperx.load_model("large-v2", device, compute_type=compute_type)

audio = whisperx.load_audio(audio_file)
result = model.transcribe(audio, batch_size=batch_size)
print(result["segments"]) # before alignment

# delete model if low on GPU resources
import gc; gc.collect(); torch.cuda.empty_cache(); del model

/usr/local/lib/python3.10/dist-packages/pyannote/audio/core/io.py:43: UserWarning: torchaudio._backend.set_audio_backend has been deprecated. With dispatcher enabled, this function is no-op. You can remove the function call.
  torchaudio.set_audio_backend("soundfile")
/usr/local/lib/python3.10/dist-packages/torch_audiomentations/utils/io.py:27: UserWarning: torchaudio._backend.set_audio_backend has been deprecated. With dispatcher enabled, this function is no-op. You can remove the function call.
  torchaudio.set_audio_backend("soundfile")


No language specified, language will be first be detected for each audio file (increases inference time).


100%|█████████████████████████████████████| 16.9M/16.9M [00:01<00:00, 10.9MiB/s]
INFO:pytorch_lightning.utilities.migration.utils:Lightning automatically upgraded your loaded checkpoint from v1.5.4 to v2.1.0. To apply the upgrade to your files permanently, run `python -m pytorch_lightning.utilities.upgrade_checkpoint ../root/.cache/torch/whisperx-vad-segmentation.bin`


Model was trained with pyannote.audio 0.0.1, yours is 3.0.1. Bad things might happen unless you revert pyannote.audio to 0.x.
Model was trained with torch 1.10.0+cu102, yours is 2.1.0+cu118. Bad things might happen unless you revert torch to 1.x.
Detected language: es (1.00) in first 30s of audio...
[{'text': ' Siempre me ha preocupado enormemente ser entendido y aceptado. Pero he aprendido que agradar o contentar a todo el mundo es imposible. Por más que hagamos, nos esforcemos o justifiquemos, siempre habrá alguien que nos busque y a quien no le guste lo que hacemos.', 'start': 0.009, 'end': 19.292}, {'text': ' En el poder de confiar en ti, comparto de la forma más clara y directa todo lo que he vivido y aprendido. No quiero guardarme nada para mí, porque sé que el cambio es posible. También puedes hacerlo tú, independientemente del lugar en el que te encuentres. Recuerda, las puertas del cambio interno están siempre abiertas para todo el que se decida pasar por ellas.', 'start': 19.

# **Traducción de la transcripción**

In [5]:
# 2. Align whisper output
model_a, metadata = whisperx.load_align_model(language_code=result["language"], device=device)
result = whisperx.align(result["segments"], model_a, metadata, audio, device, return_char_alignments=False)

print(result["segments"]) # after alignment

# delete model if low on GPU resources
# import gc; gc.collect(); torch.cuda.empty_cache(); del model_a

Downloading: "https://download.pytorch.org/torchaudio/models/wav2vec2_voxpopuli_base_10k_asr_es.pt" to /root/.cache/torch/hub/checkpoints/wav2vec2_voxpopuli_base_10k_asr_es.pt
100%|██████████| 360M/360M [00:06<00:00, 59.4MB/s]


[{'start': 0.59, 'end': 5.736, 'text': ' Siempre me ha preocupado enormemente ser entendido y aceptado.', 'words': [{'word': 'Siempre', 'start': 0.59, 'end': 0.91, 'score': 0.872}, {'word': 'me', 'start': 0.99, 'end': 1.07, 'score': 0.834}, {'word': 'ha', 'start': 1.13, 'end': 1.23, 'score': 0.822}, {'word': 'preocupado', 'start': 1.291, 'end': 1.991, 'score': 0.878}, {'word': 'enormemente', 'start': 2.071, 'end': 2.812, 'score': 0.965}, {'word': 'ser', 'start': 2.832, 'end': 3.113, 'score': 0.837}, {'word': 'entendido', 'start': 3.173, 'end': 3.753, 'score': 0.919}, {'word': 'y', 'start': 3.773, 'end': 3.794, 'score': 0.001}, {'word': 'aceptado.', 'start': 3.954, 'end': 4.594, 'score': 0.954}]}, {'start': 5.736, 'end': 11.142, 'text': 'Pero he aprendido que agradar o contentar a todo el mundo es imposible.', 'words': [{'word': 'Pero', 'start': 5.736, 'end': 5.876, 'score': 0.943}, {'word': 'he', 'start': 5.936, 'end': 6.136, 'score': 0.595}, {'word': 'aprendido', 'start': 6.156, 'end'

# **Asignar un label a cada speaker (Diarization)**

In [6]:
token = "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"

# 3. Assign speaker labels
diarize_model = whisperx.DiarizationPipeline(use_auth_token=token, device=device)

# add min/max number of speakers if known
diarize_segments = diarize_model(audio)
# diarize_model(audio, min_speakers=min_speakers, max_speakers=max_speakers)

result = whisperx.assign_word_speakers(diarize_segments, result)
#print(diarize_segments)
print(result["segments"]) # segments are now assigned speaker IDs

[{'start': 0.59, 'end': 5.736, 'text': ' Siempre me ha preocupado enormemente ser entendido y aceptado.', 'words': [{'word': 'Siempre', 'start': 0.59, 'end': 0.91, 'score': 0.872, 'speaker': 'SPEAKER_00'}, {'word': 'me', 'start': 0.99, 'end': 1.07, 'score': 0.834, 'speaker': 'SPEAKER_00'}, {'word': 'ha', 'start': 1.13, 'end': 1.23, 'score': 0.822, 'speaker': 'SPEAKER_00'}, {'word': 'preocupado', 'start': 1.291, 'end': 1.991, 'score': 0.878, 'speaker': 'SPEAKER_00'}, {'word': 'enormemente', 'start': 2.071, 'end': 2.812, 'score': 0.965, 'speaker': 'SPEAKER_00'}, {'word': 'ser', 'start': 2.832, 'end': 3.113, 'score': 0.837, 'speaker': 'SPEAKER_00'}, {'word': 'entendido', 'start': 3.173, 'end': 3.753, 'score': 0.919, 'speaker': 'SPEAKER_00'}, {'word': 'y', 'start': 3.773, 'end': 3.794, 'score': 0.001, 'speaker': 'SPEAKER_00'}, {'word': 'aceptado.', 'start': 3.954, 'end': 4.594, 'score': 0.954, 'speaker': 'SPEAKER_00'}], 'speaker': 'SPEAKER_00'}, {'start': 5.736, 'end': 11.142, 'text': 'Per

En resumen, del ultimo resultado obtenido en la parte de Diarization, lo que se debería traducir es la parte de text.

Ej:

[{'start': 0.59, 'end': 5.736, **'text': ' Siempre me ha preocupado enormemente ser entendido y aceptado.'**, 'words': [{'word': 'Siempre', 'start': 0.59, 'end': 0.91, 'score': 0.872, 'speaker': 'SPEAKER_00'}, {'word': 'me', 'start': 0.99, 'end': 1.07, 'score': 0.834, 'speaker': 'SPEAKER_00'}, {'word': 'ha', 'start': 1.13, 'end': 1.23, 'score': 0.822, 'speaker': 'SPEAKER_00'}, {'word': 'preocupado', 'start': 1.291, 'end': 1.991, 'score': 0.878, 'speaker': 'SPEAKER_00'}, {'word': 'enormemente', 'start': 2.071, 'end': 2.812, 'score': 0.965, 'speaker': 'SPEAKER_00'}, {'word': 'ser', 'start': 2.832, 'end': 3.113, 'score': 0.837, 'speaker': 'SPEAKER_00'}, {'word': 'entendido', 'start': 3.173, 'end': 3.753, 'score': 0.919, 'speaker': 'SPEAKER_00'}, {'word': 'y', 'start': 3.773, 'end': 3.794, 'score': 0.001, 'speaker': 'SPEAKER_00'}, {'word': 'aceptado.', 'start': 3.954, 'end': 4.594, 'score': 0.954, 'speaker': 'SPEAKER_00'}], 'speaker': 'SPEAKER_00'}, {'start': 5.736, 'end': 11.142, **'text': 'Pero he aprendido que agradar o contentar a todo el mundo es imposible.**', ...}]

In [7]:
for i in result["segments"]:

  a_traducir = i['text']

  print(f"{i['text']} -> start: {i['start']}, end: {i['end']}")

 Siempre me ha preocupado enormemente ser entendido y aceptado. -> start: 0.59, end: 5.736
Pero he aprendido que agradar o contentar a todo el mundo es imposible. -> start: 5.736, end: 11.142
Por más que hagamos, nos esforcemos o justifiquemos, siempre habrá alguien que nos busque y a quien no le guste lo que hacemos. -> start: 11.142, end: 18.711
 En el poder de confiar en ti, comparto de la forma más clara y directa todo lo que he vivido y aprendido. -> start: 19.672, end: 27.294
No quiero guardarme nada para mí, porque sé que el cambio es posible. -> start: 27.294, end: 32.756
También puedes hacerlo tú, independientemente del lugar en el que te encuentres. -> start: 32.756, end: 38.618
Recuerda, las puertas del cambio interno están siempre abiertas para todo el que se decida pasar por ellas. -> start: 38.618, end: 45.1
